### ISSUE TITLE
Access to KeyData.idx

### ISSUE THREAD

**Issue by BenjaminSchaaf** [orlp/slotmap#105]
I have a situation where I'm temporarily keeping some extra data for each entry in a `SlotMap` (specifically a boolean for whether or not I've seen that entry before).  I could use a `SecondaryMap`, but that requires iterating the `SlotMap` to initialize each entry, or to do extra checks each time it's checked. This part is quite performance sensitive, so ideally what I'd like to do is:

```
let mut seen = vec![];
seen.resize(slotmap.capacity(), false);
...
loop {
    if seen[key.index()] {
        continue;
```

Currently the only way to access the index is by doing `(key.data().as_ffi() & 0xFFFF) as u32`.

**Comment by orlp** [comment]
Yes, in slotmap 2.0 you will be freely able to access the index / version instead of having to go through FFI.

**Comment by muja** [comment]
Funnily this issue (which I found by looking for for the completely unrelated `initial` keyword) is similar to another use case.

It should be possible to have _some_ guarantees about keys, at least for some initial values right after map creation (e.g. static values/defaults)

I am using a slotmap as a heap which always has an initial value (i.e. the empty / default value). Instead of creating this empty value in different places, I use the initial `KeyData` value `0x100000001`.

I know this isn't specified behavior, but it works. Would this make sense to stabilize or at least provide some guarantee?
